### PR TITLE
fix: undefined escape at #836

### PR DESCRIPTION
**Pull request by atian25**
relate #843
https://github.com/mozilla/nunjucks/pull/836/files#r78286483

### CODE REVIEW

File: src/filters.js [hunk 1]
```diff
@@ -119,6 +119,7 @@ var filters = {
         if(str instanceof r.SafeString) {
             return str;
         }
+        if (str == null) str = '';
```
Comment (vecmezoni) [inline src/filters.js]: Instead of this, please use https://github.com/atian25/nunjucks/blob/d75a9993c79fb513e9e651bc938f5555213cb2a7/src/filters.js#L6

Comment (atian25) [inline src/filters.js]: `if(value === null || value === undefined || value === false) {`

`render('{{ foo }}', { foo: false })` should render `false` or `empty string` ?

**Comment by fengmk2** [comment]
@atian25 should fix `safe` filter bug too.

File: tests/compiler.js [hunk 2]
```diff
@@ -1395,6 +1395,10 @@
                 expect(res).to.be('&><\'"');
             });
 
+            render('{{ foo }}', { foo: null }, { autoescape: true }, function (err, res) {
+                expect(res).to.be('');
+            });
```
Comment (fengmk2) [inline tests/compiler.js]: add `render('{{ foo | safe }}', { foo: '<html>' }, { autoescape: true }` test for safe filter.

Comment (atian25) [inline tests/compiler.js]: done

**Comment by atian25** [comment]
@vecmezoni updated.

question: 

`if(value === null || value === undefined || value === false) {`

`render('{{ foo }}', { foo: false })` should render `false` or `empty string` ?

**Comment by fengmk2** [comment]
boolean value should follow `toString()` logic.

**Comment by atian25** [comment]
@fengmk2  agree. updated.

File: src/filters.js [hunk 3]
```diff
@@ -119,10 +119,12 @@ var filters = {
         if(str instanceof r.SafeString) {
             return str;
         }
+        str = (str === null || str === undefined) ? '' : str;
```
Comment (vecmezoni) [inline src/filters.js]: Could you please fix normalize function and remove false from there and use normalize here.
I don't see any idea behind it

Comment (atian25) [inline src/filters.js]: maybe another PR.  This one is focus at quick bug fix.
I'm afraid it will cause BREAK CHANGE.

Comment (vecmezoni) [inline src/filters.js]: ok, i'll merge it to 2.x, please prepare pull request for master

Comment (vecmezoni) [inline src/filters.js]: could you rebase this branch on 2.x?

Comment (atian25) [inline src/filters.js]: ok, wait a moment

Comment (atian25) [inline src/filters.js]: @vecmezoni plz review again. The ci is fail, but local test is ok, could you trigger it again?

Comment (vecmezoni) [inline src/filters.js]: I can't trigger CI, you can change commit hash and push it with force. It would trigger Appveyour

Comment (atian25) [inline src/filters.js]: @vecmezoni done, travis is pass.

**Comment by vecmezoni** [comment]
Merged. Waiting for `master` branch fix.

**Comment by vecmezoni** [comment]
2.5.1 released, cherry-picked this patch to master.

File: src/filters.js [hunk 4]
```diff
@@ -119,10 +119,12 @@ var filters = {
...
         return r.markSafe(lib.escape(str.toString()));
     },
 
     safe: function(str) {
+        str = (str === null || str === undefined) ? '' : str;
```
Comment (atian25) [inline src/filters.js]: @fengmk2 yes..  missing sth. https://github.com/node-modules/nunjucks/pull/3/files#diff-9f47adb6bfb92783e65552d7436a5cdcR108

**Pull request merged**